inst/www/react-tools/react-tools.js

```diff
@@ -0,0 +1,128 @@
+window.reactR = (function () {
+    /**
+     * Recursively transforms tag, a JSON representation of an instance of a
+     * React component and its children, into a React element suitable for
+     * passing to ReactDOM.render.
+     * @param {Object} components
+     * @param {Object} tag
+     */
+    function hydrate(components, tag) {
+        if (tag.name[0] !== tag.name[0].toUpperCase()) {
+            throw new Error("Component does not begin with a capital letter: " + tag.name);
+        }
+        if (!components.hasOwnProperty(tag.name)) {
+            throw new Error("Unknown component: " + tag.name);
+        }
+        var args = [components[tag.name], tag.attribs];
+        for (var i = 0; i < tag.children.length; i++) {
+            args.push(hydrate(components, tag.children[i]));
+        }
+        // Look into a way to clone an element and apply new props
+        return React.createElement.apply(null, args);
+    }
+
+    // TODO default set of options
+    // TODO Look at ReactDOM.findDOMNode()
+
+    var defaultOptions = {
+        widthProperty: "width",
+        heightProperty: "height",
+        appendPx: false,
+        renderOnResize: false
+    }
+
+    /**
+     * If options contains recognized option names, returns a new object with
+     * defaultOptions and user options merged. User options take precedence.
+     * @param {Object} options An object of user-supplied keys and values
+     * @returns {Object} The merged options
+     */
+    function mergeOptions(options) {
+        for (var k in options) {
+            if (!defaultOptions.hasOwnProperty(k)) {
+                throw new Error("Unrecognized option: " + k);
+            }
+        }
+        return jQuery.extend({}, defaultOptions, options);
```

Comment:
@alandipert I don't think we have defined a dependency on `jQuery`.  Should we or am I missing it somewhere?  Or are we just fortunate that Shiny comes with `jQuery`?

Reply:
I *think* it's safe to assume the presence of jQuery in the presence of Shiny. And I don't think this would ever run in a browser without Shiny... but that's something to ponder.